### PR TITLE
35-logging :: Added Logging functionality to project

### PR DESCRIPTION
**Pull request by RushilShaw**
Added all the specifications in the "Added Logging Issue"

### CODE REVIEW

File: src/Metrics.ts [hunk 1]
```diff
@@ -30,7 +54,11 @@ export abstract class Metrics {
     private getRepoData(url: string): { owner: string; repo: string } {
         const regex = /https:\/\/github\.com\/([^/]+)\/([^/]+)/;
         const match = url.match(regex);
-        if (!match) throw new Error("Invalid GitHub URL");
+        if (!match)
+        {
+            logger.error(`${url} is an valid Github URL`);
```
Comment (Coop8) [inline src/Metrics.ts:59]: Shouldn't this be invalid?

**Review by Coop8 — COMMENTED** [review]
Looks good, good work

File: .github/workflows/ci.yaml
Comment (galaxybomb23) [inline .github/workflows/ci.yaml:1]: Changes here will immediately be overwritten by #39 / #21. this file should not be changed. It doesn't involve "logging"

File: src/busFactor.ts [hunk 2]
```diff
@@ -73,10 +73,6 @@ export class BusFactor extends Metrics {
             page++;
         }
 
-        // print total number of commits 📝
```
Comment (galaxybomb23) [inline src/busFactor.ts:76]: the 📝 means this should be a debug level logging. NOT DELETED

File: src/Metrics.ts (hunk 1 above)
Comment (RushilShaw) [reply src/Metrics.ts:59]: Yup! I will fix push a fix for this.

File: src/cli.ts [hunk 3]
```diff
@@ -69,35 +70,37 @@ async function runTests() {
...
     // Display test results
     results.forEach((result, index) => {
         passedTests += result.passed;
         failedTests += result.failed;
     });
 
-    console.log(`\x1b[1;32mTests Passed: ${passedTests}\x1b[0m`);
-    console.log(`\x1b[1;31mTests Failed: ${failedTests}\x1b[0m`);
-    console.log('\x1b[1;34mTests complete\x1b[0m');
+    logger.info(`Tests Passed: ${passedTests}`);
```
Comment (galaxybomb23) [inline src/cli.ts:88]: i think this needs to be process.std.write()
check #39

File: src/cli.ts [hunk 4]
```diff
@@ -69,35 +70,37 @@ async function runTests() {
...
+    logger.info(`Tests Failed: ${failedTests}`);
+    logger.info("Tests complete");
 
-    // If more than 5% of the tests fail, exit with error
-    if (failedTests / (passedTests + failedTests) > 0.05) {
-        console.log('\x1b[1;31mError: More than 5% of tests failed. Exiting with error code 1.\x1b[0m');
+    if (failedTests / (passedTests + failedTests) > 0.05) {      //if more than 5% of the tests fail, exit with error
```
Comment (galaxybomb23) [inline src/cli.ts:92]: Remove this, Checker in #39 does this now

File: src/cli.ts [hunk 5]
```diff
@@ -69,35 +70,37 @@ async function runTests() {
...
+        logger.error('Error: More than 5% of tests failed. Exiting with error code 1');
+        await sleep(1000);
         exit(1);
     }
+    await sleep(1000);
```
Comment (galaxybomb23) [inline src/cli.ts:97]: slowing down the code? no?

File: src/netScore.ts [hunk 6]
```diff
@@ -61,14 +61,7 @@ export class NetScore extends Metrics {
 
         // check if netscore is between 0 and 1
         if (this.netScore < 0 || this.netScore > 1) {
-            console.log(`NetScore out of bounds: ${this.netScore}`);
-
-            // print the metrics
-            console.log(`BusFactor: ${this.busFactor.busFactor}`);
```
Comment (galaxybomb23) [inline src/netScore.ts:67]: why are we deleting this? This should be Debug? There was significant issues in calculating NetScore, so knowing what its working with (the metrics) is important for debug.

**Review by galaxybomb23 — CHANGES_REQUESTED** [review]
Lotta things here. Mainly pull from #21 since there was a lot of ouput updates there in order to meet the checker standards

**Comment by galaxybomb23** [comment]
fixed merge conflicts

File: src/cli.ts [hunk 7]
```diff
@@ -95,36 +98,35 @@ async function runTests() {
...
     let coverage: number = Math.round(passedTests / (passedTests + failedTests) * 100); // dummy variable for now
     let total: number = passedTests + failedTests;
     
-    process.stdout.write(`Total: ${total}\n`);
-    process.stdout.write(`Passed: ${passedTests}\n`);
-    process.stdout.write(`Coverage: ${coverage}%\n`);
-    process.stdout.write(`${passedTests}/${total} test cases passed. ${coverage}% line coverage achieved.\n`);
+    logger.info(`Total: ${total}`);
+    logger.info(`Passed: ${passedTests}`);
+    logger.info(`Coverage: ${coverage}%`);
```
Comment (galaxybomb23) [inline src/cli.ts:128]: why is this in info?? This is needed by the Checker... as seen in 

commit 611f8ccd2665ae08213f95a22b652fc8948a2914

**Review by Coop8 — APPROVED** [review]
Does not look good to me

File: src/Metrics.ts [hunk 8]
```diff
@@ -7,7 +8,29 @@ const githubToken = process.env.GITHUB_TOKEN;
 if (!githubToken) {
     throw new Error('API_TOKEN is not defined in the .env file');
```
Comment (galaxybomb23) [inline src/Metrics.ts:9]: GITHUB_TOKEN not API

File: src/cli.ts [hunk 9]
```diff
@@ -95,36 +98,35 @@ async function runTests() {
...
     results.forEach((result, index) => {
         passedTests += result.passed;
         failedTests += result.failed;
     });
 
-    console.log(`\x1b[1;32mTests Passed: ${passedTests}\x1b[0m`);
-    console.log(`\x1b[1;31mTests Failed: ${failedTests}\x1b[0m`);
-    console.log('\x1b[1;34mTests complete\x1b[0m');
+    process.stdout.write(`Tests Passed: ${passedTests}\n`);
+    process.stdout.write(`Tests Failed: ${failedTests}\n`);
```
Comment (galaxybomb23) [inline src/cli.ts:119]: this was removed/updated in #21 ... how is this still here?

**Review by galaxybomb23 — CHANGES_REQUESTED** [review]
There seems to be an issue where the main you pulled from was pre #39  and thus you've overwritten some of the work done in #21. not sure how This will play out

**Review by galaxybomb23 — APPROVED** [review]

**Review by Coop8 — APPROVED** [review]
Now we are talking 😎

**Comment by galaxybomb23** [comment]
finally, 
image

**Pull request merged**